app/src/org/commcare/mediadownload/MissingMediaDownloadHelper.kt

```diff
@@ -0,0 +1,196 @@
+        return REQUEST_NAME + "_" + appId
+    }
+
+
+    // Verifies all application media and re-downloads any missing lazy resources
+    fun downloadAllMissingMedia() {
+        val platform = CommCareApplication.instance().commCarePlatform
+        val global = platform.globalResourceTable
+
+        global.setInstallCancellationChecker(this)
+
+        val problems = Vector<MissingMediaException>()
+        global.verifyInstallation(problems, platform)
+
+        val missingMediaResources = SizeBoundUniqueVector<Resource>(problems.size)
+        val lazyResources = global.lazyResources
+
+        problems.filter { problem -> problem.type == MissingMediaException.MissingMediaExceptionType.FILE_NOT_FOUND }
+                .map { problem ->
+                    runCatching {
+                        missingMediaResources.add(getLazyResourceFromMediaUri(lazyResources, problem.uri))
+                    }.onFailure {
+                        Logger.exception(
+                                "Could not map to a lazy resource while downloading missing media for uri ${problem.uri}",
+                                java.lang.Exception(it))
+                    }
+                }
+
+        startPinnedNotification()
+
+        missingMediaResources.mapIndexed { index, resource ->
+            if (!wasInstallCancelled()) {
+                recoverResource(platform, resource)
+                incrementProgress(index + 1, missingMediaResources.size)
+            }
+        }
+
+        cancelNotification()
+        global.setInstallCancellationChecker(null)
+    }
+
+    /**
+     * Downloads a resource with it's location represented by {@param mediaUri}
+     */
+    @JvmStatic
+    fun requestMediaDownload(mediaUri: String, missingMediaDownloadListener: MissingMediaDownloadListener) {
+        jobs.add(
+                CoroutineScope(Dispatchers.Default).launch {
+                    var result: MissingMediaDownloadResult = MissingMediaDownloadResult.Error("Unknown Error")
+                    try {
+                        result = downloadMissingMediaResource(mediaUri)
+                    } catch (e: Exception) {
+                        Logger.exception(" An error occured while recovering a missing resource", e);
+                        withContext(Dispatchers.Main) {
+                            missingMediaDownloadListener.onComplete(MissingMediaDownloadResult.Exception(e))
+                        }
+                    }
+
+                    withContext(Dispatchers.Main) {
+                        missingMediaDownloadListener.onComplete(result)
+                    }
+                }
+        )
+    }
+
+    private fun downloadMissingMediaResource(uri: String): MissingMediaDownloadResult {
+        val platform = CommCareApplication.instance().commCarePlatform
+        val global = platform.globalResourceTable
+        val lazyResources: Vector<Resource> = global.lazyResources!!
```

Comment:
Same comment as above that I think this is likely to be too expensive to be practical if this list is as big as CAS's

Reply:
this particular call is quite cheap (takes .1 sec on MotoG, around .4 seconds on a low performant device I have) as we are [indexing the lazy property](https://github.com/dimagi/commcare-core/pull/899/files#diff-4ebf66124223ceb5a85393ee19f18ba5R335) in the Resource Table.